File: bench/graph.lua
-- The same world in Luau, with the index and the generation written by hand.
local many = 4000
local rounds = 60

local value, nextOf, generation, live = {}, {}, {}, {}
local spare = {}
local stamps = 0
local used = 0

local function add(v, n)
    local at
    if #spare > 0 then
        at = spare[#spare]
        spare[#spare] = nil
    else
        at = used
        used = used + 1
    end
    value[at] = v
    nextOf[at] = n
    stamps = stamps + 1
    generation[at] = stamps
    live[at] = true
    return generation[at] * 16777216 + at
end

local function place(handle)
    local at = handle % 16777216
    local stamp = (handle - at) / 16777216
    if not live[at] or generation[at] ~= stamp then
        return nil
    end
    return at
end

local function remove(handle)
    local at = place(handle)
    if at ~= nil then
        live[at] = false
        spare[#spare + 1] = at
    end
end

local made = {}
for i = 0, many - 1 do
    made[i] = add(i, 0)
end
for i = 0, many - 1 do
    nextOf[place(made[i])] = made[(i + 1) % many]
end

local total = 0
for round = 0, rounds - 1 do
    local at = made[0]
    for _ = 1, many do
        local slot = place(at)
        if slot == nil then return end
        total = total + value[slot]
        at = nextOf[slot]
    end
    if round % 8 == 0 then
        for i = 0, many - 1 do
            if i % 10 == round % 10 then
                local slot = place(made[i])
                if slot ~= nil then
                    local v, n = value[slot], nextOf[slot]
                    remove(made[i])
                    local again = add(v, n)
                    local before = place(made[(i + many - 1) % many])
                    if before ~= nil then
                        nextOf[before] = again
                    end
                    made[i] = again
                end
            end
        end
    end
end

print(string.format("graph %d", total))

File: bench/graph.cpp
// The same world without a language that has checked identity in it, which is
// an index and a generation beside it and a check at every read -- written by
// hand, because that is what a program without `store` and `ref` does.
#include <cstdio>
#include <cstdint>
#include <vector>

struct Node {
    int value;
    int64_t next;
};

struct World {
    std::vector<Node> slots;
    std::vector<uint32_t> generation;
    std::vector<bool> live;
    std::vector<uint32_t> spare;
    uint32_t stamps = 0;

    int64_t add(Node one) {
        uint32_t at;
        if (!spare.empty()) {
            at = spare.back();
            spare.pop_back();
        } else {
            at = (uint32_t)slots.size();
            slots.push_back({});
            generation.push_back(0);
            live.push_back(false);
        }
        slots[at] = one;
        generation[at] = ++stamps;
        live[at] = true;
        return ((int64_t)generation[at] << 24) | (int64_t)at;
    }

    Node *get(int64_t handle) {
        uint32_t at = (uint32_t)(handle & 0xffffff);
        uint32_t stamp = (uint32_t)(handle >> 24);
        if (at >= slots.size() || !live[at] || generation[at] != stamp) {
            return nullptr;
        }
        return &slots[at];
    }

    void remove(int64_t handle) {
        uint32_t at = (uint32_t)(handle & 0xffffff);
        if (get(handle) != nullptr) {
            live[at] = false;
            spare.push_back(at);
        }
    }
};

int main(void) {
    const int many = 4000;
    const int rounds = 60;
    World world;
    std::vector<int64_t> made;
    made.reserve(many);
    for (int i = 0; i < many; i++) {
        made.push_back(world.add({i, 0}));
    }
    for (int i = 0; i < many; i++) {
        Node *one = world.get(made[i]);
        one->next = made[(i + 1) % many];
    }

    long long total = 0;
    for (int round = 0; round < rounds; round++) {
        int64_t at = made[0];
        for (int steps = 0; steps < many; steps++) {
            Node *one = world.get(at);
            if (one == nullptr) {
                return 1;
            }
            total += one->value;
            at = one->next;
        }
        if (round % 8 == 0) {
            for (int i = 0; i < many; i++) {
                if (i % 10 != round % 10) {
                    continue;
                }
                Node *one = world.get(made[i]);
                if (one == nullptr) {
                    continue;
                }
                int value = one->value;
                int64_t after = one->next;
                world.remove(made[i]);
                int64_t again = world.add({value, after});
                int before = (i + many - 1) % many;
                Node *holder = world.get(made[before]);
                if (holder != nullptr) {
                    holder->next = again;
                }
                made[i] = again;
            }
        }
    }
    printf("graph %lld\n", total);
    return 0;
}
